README.md
# AnyEvent::WebSocket::Client ![linux](https://github.com/uperl/AnyEvent-WebSocket-Client/workflows/linux/badge.svg) ![windows](https://github.com/uperl/AnyEvent-WebSocket-Client/workflows/windows/badge.svg) ![macos](https://github.com/uperl/AnyEvent-WebSocket-Client/workflows/macos/badge.svg) ![msys2-mingw](https://github.com/uperl/AnyEvent-WebSocket-Client/workflows/msys2-mingw/badge.svg)

WebSocket client for AnyEvent

# SYNOPSIS

```perl
use AnyEvent::WebSocket::Client 0.12;

my $client = AnyEvent::WebSocket::Client->new;

$client->connect("ws://localhost:1234/service")->cb(sub {

  # make $connection an our variable rather than
  # my so that it will stick around.  Once the
  # connection falls out of scope any callbacks
  # tied to it will be destroyed.
  our $connection = eval { shift->recv };
  if($@) {
    # handle error...
    warn $@;
    return;
  }

  # send a message through the websocket...
  $connection->send('a message');

  # recieve message from the websocket...
  $connection->on(each_message => sub {
    # $connection is the same connection object
    # $message isa AnyEvent::WebSocket::Message
    my($connection, $message) = @_;
    ...
  });

  # handle a closed connection...
  $connection->on(finish => sub {
    # $connection is the same connection object
    my($connection) = @_;
    ...
  });

  # close the connection (either inside or
  # outside another callback)
  $connection->close;

});

## uncomment to enter the event loop before exiting.
## Note that calling recv on a condition variable before
## it has been triggered does not work on all event loops
#AnyEvent->condvar->recv;
```

# DESCRIPTION

This class provides an interface to interact with a web server that provides
services via the WebSocket protocol in an [AnyEvent](https://metacpan.org/pod/AnyEvent) context.  It uses
[Protocol::WebSocket](https://metacpan.org/pod/Protocol::WebSocket) rather than reinventing the wheel.  You could use
[AnyEvent](https://metacpan.org/pod/AnyEvent) and [Protocol::WebSocket](https://metacpan.org/pod/Protocol::WebSocket) directly if you wanted finer grain
control, but if that is not necessary then this class may save you some time.

The recommended API was added to the [AnyEvent::WebSocket::Connection](https://metacpan.org/pod/AnyEvent::WebSocket::Connection)
class with version 0.12, so it is recommended that you include that version
when using this module.  The older version of the API has since been
deprecated and removed.

# ATTRIBUTES

## timeout

Timeout for the initial connection to the web server.  The default
is 30.

## ssl\_no\_verify

If set to true, then secure WebSockets (those that use SSL/TLS) will
not be verified.  The default is false.

## ssl\_ca\_file

Provide your own CA certificates file instead of using the system default for
SSL/TLS verification.

## protocol\_version

The protocol version.  See [Protocol::WebSocket](https://metacpan.org/pod/Protocol::WebSocket) for the list of supported
WebSocket protocol versions.

## subprotocol

List of subprotocols to request from the server.  This class will throw an
exception if none of the protocols are supported by the server.

## http\_headers

Extra headers to include in the initial request.  May be either specified
as a hash reference, or an array reference.  For example:

```perl
AnyEvent::WebSocket::Client->new(
  http_headers => {
    'X-Foo' => 'bar',
    'X-Baz' => [ 'abc', 'def' ],
  },
);

AnyEvent::WebSocket::Client->new(
  http_headers => [
    'X-Foo' => 'bar',
    'X-Baz' => 'abc',
    'X-Baz' => 'def',
  ],
);
```

Will generate:

```
X-Foo: bar
X-Baz: abc
X-Baz: def
```

Although, the order cannot be guaranteed when using the hash style.

## max\_payload\_size

The maximum payload size for received frames.  Currently defaults to whatever
[Protocol::WebSocket](https://metacpan.org/pod/Protocol::WebSocket) defaults to.

## max\_fragments

The maximum number of fragments for received frames.  Currently defaults to whatever
[Protocol::WebSocket](https://metacpan.org/pod/Protocol::WebSocket) defaults to.

## env\_proxy

If you set true to this boolean attribute, it loads proxy settings
from environment variables. If it finds valid proxy settings,
`connect` method will use that proxy.

Default: false.

For `ws` WebSocket end-points, first it reads `ws_proxy` (or
`WS_PROXY`) environment variable. If it is not set or empty string,
then it reads `http_proxy` (or `HTTP_PROXY`). For `wss` WebSocket
end-points, it reads `wss_proxy` (`WSS_PROXY`) and `https_proxy`
(`HTTPS_PROXY`) environment variables.

# METHODS

## connect

```perl
my $cv = $client->connect($uri)
my $cv = $client->connect($uri, $host, $port);
```

Open a connection to the web server and open a WebSocket to the resource
defined by the given URL.  The URL may be either an instance of [URI::ws](https://metacpan.org/pod/URI::ws),
[URI::wss](https://metacpan.org/pod/URI::wss), or a string that represents a legal WebSocket URL.

You can  override the connection host and port by passing them in as the
second and third argument.  These values (if provided) are passed directly
into [AnyEvent::Socket](https://metacpan.org/pod/AnyEvent::Socket)'s `tcp_connect` function, so please note that
function's idiosyncrasies in the [AnyEvent::Socket](https://metacpan.org/pod/AnyEvent::Socket) documentation.  In
particular,  you can pass in `unix/` as the host and a filesystem path
as the "port" to connect to a unix domain socket.

This method will return an [AnyEvent](https://metacpan.org/pod/AnyEvent) condition variable which you can
attach a callback to.  The value sent through the condition variable will
be either an instance of [AnyEvent::WebSocket::Connection](https://metacpan.org/pod/AnyEvent::WebSocket::Connection) or a croak
message indicating a failure.  The synopsis above shows how to catch
such errors using `eval`.

# FAQ

## My program exits before doing anything, what is up with that?

See this FAQ from [AnyEvent](https://metacpan.org/pod/AnyEvent):
[AnyEvent::FAQ#My-program-exits-before-doing-anything-whats-going-on](https://metacpan.org/pod/AnyEvent::FAQ#My-program-exits-before-doing-anything-whats-going-on).

It is probably also a good idea to review the [AnyEvent](https://metacpan.org/pod/AnyEvent) documentation
if you are new to [AnyEvent](https://metacpan.org/pod/AnyEvent) or event-based programming.

## My callbacks aren't being called!

Make sure that the connection object is still in scope.  This often happens
if you use a `my $connection` variable and don't save it somewhere.  For
example:

```perl
$client->connect("ws://foo/service")->cb(sub {

  my $connection = eval { shift->recv };

  if($@)
  {
    warn $@;
    return;
  }

  ...
});
```

Unless `$connection` is saved somewhere it will get deallocated along with
any associated message callbacks will also get deallocated once the connect
callback is executed.  One way to make sure that the connection doesn't
get deallocated is to make it a `our` variable (as in the synopsis above)
instead.

# CAVEATS

This is pretty simple minded and there are probably WebSocket features
that you might like to use that aren't supported by this distribution.
Patches are encouraged to improve it.

# SEE ALSO

- [AnyEvent::WebSocket::Connection](https://metacpan.org/pod/AnyEvent::WebSocket::Connection)
- [AnyEvent::WebSocket::Message](https://metacpan.org/pod/AnyEvent::WebSocket::Message)
- [AnyEvent::WebSocket::Server](https://metacpan.org/pod/AnyEvent::WebSocket::Server)
- [AnyEvent](https://metacpan.org/pod/AnyEvent)
- [URI::ws](https://metacpan.org/pod/URI::ws)
- [URI::wss](https://metacpan.org/pod/URI::wss)
- [Protocol::WebSocket](https://metacpan.org/pod/Protocol::WebSocket)
- [Net::WebSocket::Server](https://metacpan.org/pod/Net::WebSocket::Server)
- [Net::Async::WebSocket](https://metacpan.org/pod/Net::Async::WebSocket)
- [RFC 6455 The WebSocket Protocol](http://tools.ietf.org/html/rfc6455)

# AUTHOR

Author: Graham Ollis <plicease@cpan.org>

Contributors:

Toshio Ito (debug-ito, TOSHIOITO)

José Joaquín Atria (JJATRIA)

Kivanc Yazan (KYZN)

Yanick Champoux (YANICK)

Fayland Lam (FAYLAND)

Daniel Kamil Kozar (xavery)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013-2022 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.55      2023-05-04 10:28:24 -0600
  [WARNING: change in behavior]
  - Fix the default code for close to be 1000 (gh#61, gh#62)

0.54      2021-03-19 06:13:48 -0600
  - Delay loading AnyEvent::Connector until needed.
  - Moved git repository to new github org.

0.53      2019-04-01 08:42:05 -0400
  - Add missing timeout from test that could otherwise hang with a
    broken combination of Debian Buster / Net::SSLeay / AnyEvent.
    (gh#46. gh#47)

0.52      2019-02-08 14:22:11 -0500
  - Production version functionally identical to the 0.51_01 release.

0.51_01   2019-02-05 23:22:05 -0500
  - You may now overide the host and port when using the
    AnyEvent::WebSocket::Client's connect method.

0.50      2019-01-14 10:49:47 -0500
  - Fixed a testing bug that was reportd on Debian Buster.

0.49      2018-07-09 21:06:24 -0400
  - Fixed a bug where adding next_message callback within a next_message
    callback was called immediately with the same message (gh#39 xavery++)

0.48      2018-05-21 21:08:49 -0400
  - Production version functionally identical to the 0.47_01 release.

0.47_01   2018-05-20 09:35:23 -0400
  - Add max_fragments attribute ot AnyEvent::WebSocket::Client
  - Add max_fragments attribute ot AnyEvent::WebSocket::Connection

0.46      2018-04-07 16:39:29 -0400
  - Production version functionally identical to the 0.45_01 release.

0.45_01   2018-04-04 17:01:03 -0400
  - add env_proxy property to AnyEvent::WebSocket::Client (gh#36 gh#37 TOSHIOITO++)

0.44      2017-12-12 15:37:51 -0500
  - Compatibility with 0.22.

0.43      2017-10-19 17:06:51 -0400
  - Production version functionally identical to the 0.42_01 release.

0.42_01   2017-10-18 11:45:16 -0400
  - You can now unregister event callbacks.
  - The 'on' method for AnyEvent::WebSocket::Connection now returns a
    code ref that can be used to unregister the callback.
  - The 'each_message' event also gets a code ref that can be used to
    unregister the callback.

  [WARNING: change in behavior]
  - The 'on' method for AnyEvent::WebSocket::Connection used to return
    the connection object, the intent was to allow chaining of 'on' calls,
    but this was never documented, and I don't believe it was ever used.
  - The 'each_message' event has a different number of arguments, which
    could break code that is using 'pop' to get the message instead of
    using the second argument position.  If you were following the examples
    in the documentation you wouldn't have done this.

0.41      2017-08-11 14:05:40 -0400
  - Production version functionally identical to the 0.40_03 release.  

0.40_03   2017-08-10 14:53:41 -0400
  - Possible workaround for Perl 5.8.x bug (gh#19)

0.40_02   2017-08-10 12:39:28 -0400
  - Fix testing testing failure introduced in 0.40_01

0.40_01   2017-08-06 22:59:50 -0400
  - Add close_code, close_reason and close_error properties to AnyEvent::Connection
  - Significant refactor of the test suite that reduces dependencies on Mojolicious
  - Migrate to MakeMaker for installer

0.39      2017-07-07 18:32:02 -0400
  - Require AnyEvent 7.13 or better to address SSL/TLS error (gh#22)
  - Updated example to use Mojo::Redis2 and modern Mojolicious (Kivanc Yazan gh#29)

0.38      2017-01-31 13:11:24 -0500
  - Fix skip in t/mojo_close_codes.t which requires EV to be installed (José Joaquín Atria gh#28)

0.37      2016-11-04 22:24:11 -0400
  - Added max_payload_size to AnyEvent::WebSocket::Client and 
    AnyEvent::WebSocket::Connection for limiting the size of
    received frames
  - Sending payloads larger than 65536 no longer causes an exception
  - Added parse_error to AnyEvent::WebSocket::Connection

0.36      2016-10-27 11:57:12 -0400
  - Production version functionally identical to the 0.35_02 release.

0.35_02   2016-10-25 13:04:54 -0400
  - Added http_headers attribute which allows you to provide additional
    arbitrary headers

0.35_01   2016-10-24 18:28:00 -0400
  - You may now specify the protocol version using the protocol_version
    attribute for AnyEvent::WebSocket::Client.  All protocol versions
    supported by Protocol::WebSocket should work.
  - Added subprotocol attribute for AnyEvent::WebSocket::Client

0.35      2016-08-04 13:47:36 -0400
  - Production version functionally identical to the 0.34_01 release.

0.34_01   2016-07-31 08:42:40 -0400
  - Fixed bug where read data could be discarded when in TLS mode.
    (gh#25 thanks Toshio Ito)

0.34      2016-07-26 14:04:52 -0400
  - Documentation improvements
  - Production version functionally identical to the 0.33_01 release.

0.33_01   2016-07-25 22:48:06 -0400
  - Added optional code and reason arguments to $connection->close (gh#23 thanks José Joaquín Atria)

0.33      2016-04-25 12:12:12 -0400
  - t/ae_ssl.t may now be skipped by setting ANYEVENT_WEBSOCKET_TEST_SKIP_SSL

0.32      2015-12-17 14:02:36 -0500
  - URI 1.53 or better has been required for a while (if not since 0.01)
    This is now reflected in the module metadata.
    Thanks to Fabien Wernli (faxm0dem) for the report.

0.31      2015-10-11 06:21:29 -0400
  - Removed deprecated methods for AnyEvent::WebSocket::Connection class:
    on_each_message
    on_next_message
    on_finish

0.30      2015-03-03 23:33:03 -0500
  - Require Moo 2.0, as older versions inadvertently turn on
    fatal warnings

0.29      2015-01-17 10:03:16 -0500
  - document when deprecated methods will be removed.

0.28      2014-12-08 12:43:53 -0500
  - make confusing documentation less confusing.

0.27      2014-08-20 03:54:14 -0400
  - disable t/mojo_ssl.t since it is unreliable and we test the 
    same features with t/ae_ssl.t now.

0.26      2014-08-19 10:59:26 -0400
  - if Crypt::Random::Source is already installed, require at least
    version 0.08 to avoid Class::MOP deprecation warnings

0.25      2014-08-11 14:35:11 -0400
  - fix broken link in documentation

0.24      2014-08-11 07:29:49 -0400
  - documentation

t/testlib/Mojo.pm
package testlib::Mojo;

use strict;
use warnings;
use Mojo::Server::Daemon;
use Test::More;

sub start_mojo
{
  my ($class, %args) = @_;
  my $app = $args{app};
  my $scheme = $args{ssl} ? "https" : "http";
  my $server = Mojo::Server::Daemon->new;
  my $port = $server->ioloop->generate_port;
  note "port = $port";
  $server->app($app);
  $server->listen(["$scheme://127.0.0.1:$port"]);
  $server->start;
  return ($server, $port);
}


1;


t/testlib/Server.pm
package testlib::Server;

use strict;
use warnings;
use URI;
use Test::More;
use AnyEvent::Handle;
use AnyEvent::Socket qw( tcp_server);
use Protocol::WebSocket::Handshake::Server;
use Protocol::WebSocket::Frame;
use PerlX::Maybe qw( maybe );

my $timeout;

sub set_timeout
{
  return if defined $timeout;
  $timeout = AnyEvent->timer( after => 30, cb => sub {
    diag "timeout!";
    exit 2;
  });
}

sub start_server
{
  my $class = shift;
  my $opt = { @_ };
  $opt->{handshake} ||= sub {};
  $opt->{customize_server_response} ||= sub {};
  my $server_cv = AnyEvent->condvar;

  tcp_server undef, undef, sub {
    my $handshake = Protocol::WebSocket::Handshake::Server->new;
    my $frame     = Protocol::WebSocket::Frame->new( max_payload_size => 9223372036854775807 );
  
    my $hdl = AnyEvent::Handle->new(
      $opt->{tls} ? (tls => 'accept', tls_ctx => $opt->{tls}) : (),
      fh => shift,
      on_eof => sub {
        note("testlib::Server: on_eof called.");
      }
    );
  
    $hdl->on_read(
      sub {
        my $chunk = $_[0]{rbuf};
        $_[0]{rbuf} = '';

        unless($handshake->is_done) {
          $handshake->parse($chunk);
          if($handshake->is_done)
          {
            $opt->{customize_server_response}->($handshake);
            $hdl->push_write($handshake->to_string);
            $opt->{handshake}->(handshake => $handshake, hdl => $hdl);
          }
          return;
        }
      
        $frame->append($chunk);
      
        while(defined(my $message = $frame->next))
        {
          $opt->{message}->(frame => $frame, message => $message, hdl => $hdl);
        }
      }
    );
  }, sub {
    my($fh, $host, $port) = @_;
    $server_cv->send($port);
  };

  my $port = $server_cv->recv;
  
  my $uri = URI->new('ws://127.0.0.1/echo');
  $uri->port($port);
  $uri->scheme('wss') if $opt->{tls};
  note "$uri";
  $uri;
}

sub start_echo
{
  shift->start_server(message => sub {
    my $opt = { @_ };
    
    return if !$opt->{frame}->is_text && !$opt->{frame}->is_binary;

    
    $opt->{hdl}->push_write($opt->{frame}->new(buffer => $opt->{message}, max_payload_size => 9223372036854775807 )->to_bytes);
        
    if($opt->{message} eq 'quit')
    {
      $opt->{hdl}->push_write($opt->{frame}->new(type => 'close')->to_bytes);
      $opt->{hdl}->push_shutdown;
    }
  });
}

1;


t/testlib/Connection.pm
package testlib::Connection;
use strict;
use warnings;
use AnyEvent::WebSocket::Connection;
use AnyEvent::Handle;
use AnyEvent::Socket qw(tcp_server);

sub create_handle_pair
{
  my $cv_port = AnyEvent->condvar;
  my $cv_server_fh = AnyEvent->condvar;
  my $server = tcp_server undef, undef, sub {
    my ($fh) = @_;
    $cv_server_fh->send($fh);
  }, sub {
    my($fh, $host, $port) = @_;
    $cv_port->send($port);
  };
  my $cv_connect = AnyEvent->condvar;
  my $a_handle = AnyEvent::Handle->new(
    connect => ["127.0.0.1", $cv_port->recv],
    on_error => sub { die "connect error: $_[2]" },
    on_connect => sub { $cv_connect->send() }
  );
  $cv_connect->recv;
  my $b_handle = AnyEvent::Handle->new(
     fh => $cv_server_fh->recv  
  );
  return ($a_handle, $b_handle);
}

sub create_connection_pair
{
  my ($class, $a_options_ref, $b_options_ref) = @_;
  $a_options_ref ||= {};
  $b_options_ref ||= {};
  my ($a_handle, $b_handle) = $class->create_handle_pair();
  return (
    AnyEvent::WebSocket::Connection->new(%$a_options_ref, handle => $a_handle),
    AnyEvent::WebSocket::Connection->new(%$b_options_ref, handle => $b_handle),
  );
}

sub create_connection_and_handle
{
  my ($class, $a_options_ref) = @_;
  my ($a_handle, $b_handle) = $class->create_handle_pair();
  return (
    AnyEvent::WebSocket::Connection->new(%$a_options_ref, handle => $a_handle),
    $b_handle
  );
}

1;


lib/AnyEvent/WebSocket/Message.pm
package AnyEvent::WebSocket::Message;

use strict;
use warnings;
use Moo;
use Encode ();

# ABSTRACT: WebSocket message for AnyEvent
# VERSION

=head1 SYNOPSIS

 $connection->send(
   AnyEvent::WebSocket::Message->new(body => "some message"),
 );
 
 $connection->on(each_message => sub {
   my($connection, $message) = @_;
   if($message->is_text || $message->is_binary)
   {
     my $body = $message->body;
   }
 });

=head1 DESCRIPTION

Instances of this class represent a single WebSocket message.  They are
the objects that come through from the other end of your
L<AnyEvent::WebSocket::Connection> instance.  They can also be sent through
that class using its C<send> method.

=head1 ATTRIBUTES

=head2 body

The body or payload of the message.

=head2 opcode

The integer code for the type of message.

=cut

has body => ( is => 'ro', required => 1 );
has opcode => ( is => 'ro', default => 1 );

=head1 METHODS

=head2 decoded_body

 my $body = $message->decoded_body;

Returns the body decoded from UTF-8.

=cut

sub decoded_body
{
  Encode::decode("UTF-8", shift->body)
}

=head2 is_text

 my $bool = $message->is_text;

True if the message is text.

=head2 is_binary

 my $bool = $message->is_binary;

True if the message is binary.

=head2 is_close

 my $bool = $message->is_close;

True if the message is a close message.

=head2 is_ping

 my $bool = $message->is_ping

True if the message is a ping.

=head2 is_pong

 my $bool = $message->is_pong;

True if the message is a pong.

=cut

sub is_text   { $_[0]->opcode == 1 }
sub is_binary { $_[0]->opcode == 2 }
sub is_close  { $_[0]->opcode == 8 }
sub is_ping   { $_[0]->opcode == 9 }
sub is_pong   { $_[0]->opcode == 10 }

1;

=head1 SEE ALSO

=over 4

=item *

L<AnyEvent::WebSocket::Client>

=item *

L<AnyEvent::WebSocket::Connection>

=item *

L<AnyEvent::WebSocket::Server>

=item *

L<AnyEvent>

=item *

L<RFC 6455 The WebSocket Protocol|http://tools.ietf.org/html/rfc6455>

=back

=begin stopwords

Joaquín José

=end stopwords

=cut


lib/AnyEvent/WebSocket/Client.pm
package AnyEvent::WebSocket::Client;

use strict;
use warnings;
use Moo;
use AE;
use AnyEvent;
use AnyEvent::Handle;
use AnyEvent::Socket ();
use Protocol::WebSocket::Request;
use Protocol::WebSocket::Handshake::Client;
use AnyEvent::WebSocket::Connection;
use PerlX::Maybe qw( maybe provided );

# ABSTRACT: WebSocket client for AnyEvent
# VERSION

=head1 SYNOPSIS

 use AnyEvent::WebSocket::Client 0.12;
 
 my $client = AnyEvent::WebSocket::Client->new;
 
 $client->connect("ws://localhost:1234/service")->cb(sub {
 
   # make $connection an our variable rather than
   # my so that it will stick around.  Once the
   # connection falls out of scope any callbacks
   # tied to it will be destroyed.
   our $connection = eval { shift->recv };
   if($@) {
     # handle error...
     warn $@;
     return;
   }
 
   # send a message through the websocket...
   $connection->send('a message');
 
   # recieve message from the websocket...
   $connection->on(each_message => sub {
     # $connection is the same connection object
     # $message isa AnyEvent::WebSocket::Message
     my($connection, $message) = @_;
     ...
   });
 
   # handle a closed connection...
   $connection->on(finish => sub {
     # $connection is the same connection object
     my($connection) = @_;
     ...
   });
 
   # close the connection (either inside or
   # outside another callback)
   $connection->close;
 
 });
 
 ## uncomment to enter the event loop before exiting.
 ## Note that calling recv on a condition variable before
 ## it has been triggered does not work on all event loops
 #AnyEvent->condvar->recv;

=head1 DESCRIPTION

This class provides an interface to interact with a web server that provides
services via the WebSocket protocol in an L<AnyEvent> context.  It uses
L<Protocol::WebSocket> rather than reinventing the wheel.  You could use
L<AnyEvent> and L<Protocol::WebSocket> directly if you wanted finer grain
control, but if that is not necessary then this class may save you some time.

The recommended API was added to the L<AnyEvent::WebSocket::Connection>
class with version 0.12, so it is recommended that you include that version
when using this module.  The older version of the API has since been
deprecated and removed.

=head1 ATTRIBUTES

=head2 timeout

Timeout for the initial connection to the web server.  The default
is 30.

=cut

has timeout => (
  is      => 'ro',
  default => sub { 30 },
);

=head2 ssl_no_verify

If set to true, then secure WebSockets (those that use SSL/TLS) will
not be verified.  The default is false.

=cut

has ssl_no_verify => (
  is => 'ro',
);

=head2 ssl_ca_file

Provide your own CA certificates file instead of using the system default for
SSL/TLS verification.

=cut

has ssl_ca_file => (
  is => 'ro',
);

=head2 protocol_version

The protocol version.  See L<Protocol::WebSocket> for the list of supported
WebSocket protocol versions.

=cut

has protocol_version => (
  is => 'ro',
);

=head2 subprotocol

List of subprotocols to request from the server.  This class will throw an
exception if none of the protocols are supported by the server.

=cut

has subprotocol => (
  is     => 'ro',
  coerce => sub { ref $_[0] ? $_[0] : [$_[0]] },
);

=head2 http_headers

Extra headers to include in the initial request.  May be either specified
as a hash reference, or an array reference.  For example:

 AnyEvent::WebSocket::Client->new(
   http_headers => {
     'X-Foo' => 'bar',
     'X-Baz' => [ 'abc', 'def' ],
   },
 );
 
 AnyEvent::WebSocket::Client->new(
   http_headers => [
     'X-Foo' => 'bar',
     'X-Baz' => 'abc',
     'X-Baz' => 'def',
   ],
 );

Will generate:

 X-Foo: bar
 X-Baz: abc
 X-Baz: def

Although, the order cannot be guaranteed when using the hash style.

=cut

has http_headers => (
  is => 'ro',
  coerce => sub {
    ref $_[0] eq 'ARRAY' ? $_[0] : do {
      my $h = shift;
      [
        map {
          my($k,$v) = ($_, $h->{$_});
          $v = [$v] unless ref $v;
          map { $k => $_ } @$v;
          # sorted to make testing easier.
          # may be removed in the future
          # so do not depend on it.
        } sort keys %$h
      ],
    };
  },
);

=head2 max_payload_size

The maximum payload size for received frames.  Currently defaults to whatever
L<Protocol::WebSocket> defaults to.

=cut

has max_payload_size => (
  is => 'ro',
);

=head2 max_fragments

The maximum number of fragments for received frames.  Currently defaults to whatever
L<Protocol::WebSocket> defaults to.

=cut

has max_fragments => (
  is => 'ro',
);

=head2 env_proxy

If you set true to this boolean attribute, it loads proxy settings
from environment variables. If it finds valid proxy settings,
C<connect> method will use that proxy.

Default: false.

For C<ws> WebSocket end-points, first it reads C<ws_proxy> (or
C<WS_PROXY>) environment variable. If it is not set or empty string,
then it reads C<http_proxy> (or C<HTTP_PROXY>). For C<wss> WebSocket
end-points, it reads C<wss_proxy> (C<WSS_PROXY>) and C<https_proxy>
(C<HTTPS_PROXY>) environment variables.

=cut

has env_proxy => (
  is => 'ro',
  default => sub { 0 },
);


=head1 METHODS

=head2 connect

 my $cv = $client->connect($uri)
 my $cv = $client->connect($uri, $host, $port);

Open a connection to the web server and open a WebSocket to the resource
defined by the given URL.  The URL may be either an instance of L<URI::ws>,
L<URI::wss>, or a string that represents a legal WebSocket URL.

You can  override the connection host and port by passing them in as the
second and third argument.  These values (if provided) are passed directly
into L<AnyEvent::Socket>'s C<tcp_connect> function, so please note that
function's idiosyncrasies in the L<AnyEvent::Socket> documentation.  In
particular,  you can pass in C<unix/> as the host and a filesystem path
as the "port" to connect to a unix domain socket.

This method will return an L<AnyEvent> condition variable which you can
attach a callback to.  The value sent through the condition variable will
be either an instance of L<AnyEvent::WebSocket::Connection> or a croak
message indicating a failure.  The synopsis above shows how to catch
such errors using C<eval>.

=cut

sub connect
{
  my($self, $uri, $host, $port) = @_;
  unless(ref $uri)
  {
    require URI;
    $uri = URI->new($uri);
  }

  my $done = AE::cv;

  # TODO: should we also accept http and https URLs?
  # probably.
  if($uri->scheme ne 'ws' && $uri->scheme ne 'wss')
  {
    $done->croak("URI is not a websocket");
    return $done;
  }

  $host = $uri->host unless defined $host;
  $port = $uri->port unless defined $port;

  $self->_make_tcp_connection($uri->scheme, $host, $port, sub {
    my $fh = shift;
    unless($fh)
    {
      $done->croak("unable to connect");
      return;
    }
    my $req = Protocol::WebSocket::Request->new( maybe headers => $self->http_headers );
    my $handshake = Protocol::WebSocket::Handshake::Client->new(
            url     => $uri->as_string,
      maybe version => $self->protocol_version,
            req     => $req,
    );

    my %subprotocol;
    if($self->subprotocol)
    {
      %subprotocol = map { $_ => 1 } @{ $self->subprotocol };
      $handshake->req->subprotocol(join(',', @{ $self->subprotocol }));
    }

    my $hdl = AnyEvent::Handle->new(
                                                      fh       => $fh,
      provided $uri->secure,                          tls      => 'connect',
      provided $uri->secure && !$self->ssl_no_verify, peername => $uri->host,
      provided $uri->secure && !$self->ssl_no_verify, tls_ctx  => {
                                                              verify => 1,
                                                              verify_peername => "https",
                                                        maybe ca_file => $self->ssl_ca_file,
                                                      },
                                                      on_error => sub {
                                                        my ($hdl, $fatal, $msg) = @_;
                                                        if($fatal)
                                                        { $done->croak("connect error: " . $msg) }
                                                        else
                                                        { warn $msg }
                                                      },
    );

    $hdl->push_write($handshake->to_string);
    $hdl->on_read(sub {
      $handshake->parse($_[0]{rbuf});
      if($handshake->error)
      {
        $done->croak("handshake error: " . $handshake->error);
        undef $hdl;
        undef $handshake;
        undef $done;
      }
      elsif($handshake->is_done)
      {
        my $sb;
        if($self->subprotocol)
        {
          $sb = $handshake->res->subprotocol;
          if(defined $sb)
          {
            unless($subprotocol{$sb})
            {
              $done->croak("subprotocol mismatch, requested: @{[ join ', ', @{ $self->subprotocol } ]}, got: $sb");
            }
          }
          else
          {
            $done->croak("no subprotocol in response");
          }
        }
        undef $handshake;
        $done->send(
          AnyEvent::WebSocket::Connection->new(
                  handle               => $hdl,
                  masked               => 1,
            maybe subprotocol          => $sb,
            maybe max_payload_size     => $self->max_payload_size,
            maybe max_fragments        => $self->max_fragments,
          )
        );
        undef $hdl;
        undef $done;
      }
    });
  }, sub { $self->timeout });
  $done;
}

sub _make_tcp_connection
{
  my $self = shift;
  my $scheme = shift;
  my ($host, $port) = @_;
  if(!$self->env_proxy)
  {
    return &AnyEvent::Socket::tcp_connect(@_);
  }
  require AnyEvent::Connector;
  my @connectors =
      $scheme eq "ws"
      ? (map { AnyEvent::Connector->new(env_proxy => $_) } qw(ws http))
      : $scheme eq "wss"
      ? (map { AnyEvent::Connector->new(env_proxy => $_) } qw(wss https))
      : ();
  foreach my $connector (@connectors)
  {
    if(defined($connector->proxy_for($host, $port)))
    {
      return $connector->tcp_connect(@_);
    }
  }
  return &AnyEvent::Socket::tcp_connect(@_);
}

1;

=head1 FAQ

=head2 My program exits before doing anything, what is up with that?

See this FAQ from L<AnyEvent>:
L<AnyEvent::FAQ#My-program-exits-before-doing-anything-whats-going-on>.

It is probably also a good idea to review the L<AnyEvent> documentation
if you are new to L<AnyEvent> or event-based programming.

=head2 My callbacks aren't being called!

Make sure that the connection object is still in scope.  This often happens
if you use a C<my $connection> variable and don't save it somewhere.  For
example:

 $client->connect("ws://foo/service")->cb(sub {
 
   my $connection = eval { shift->recv };
 
   if($@)
   {
     warn $@;
     return;
   }
 
   ...
 });

Unless C<$connection> is saved somewhere it will get deallocated along with
any associated message callbacks will also get deallocated once the connect
callback is executed.  One way to make sure that the connection doesn't
get deallocated is to make it a C<our> variable (as in the synopsis above)
instead.

=head1 CAVEATS

This is pretty simple minded and there are probably WebSocket features
that you might like to use that aren't supported by this distribution.
Patches are encouraged to improve it.

=head1 SEE ALSO

=over 4

=item *

L<AnyEvent::WebSocket::Connection>

=item *

L<AnyEvent::WebSocket::Message>

=item *

L<AnyEvent::WebSocket::Server>

=item *

L<AnyEvent>

=item *

L<URI::ws>

=item *

L<URI::wss>

=item *

L<Protocol::WebSocket>

=item *

L<Net::WebSocket::Server>

=item *

L<Net::Async::WebSocket>

=item *

L<RFC 6455 The WebSocket Protocol|http://tools.ietf.org/html/rfc6455>

=back

=begin stopwords

Joaquín José

=end stopwords

=cut
